[Metrics] Model FLOPs Utilization estimation

SungMinCho opened the pull request:
Signed-off-by: SungMinCho <tjdals4565@gmail.com>

## TL;DR

This PR implements optional "MFU stats logging", which appends "per-GPU flops/bandwidth" information to the existing periodic logs, reporting the average compute/memory performance of GPUs in each Engine for the duration of that log interval. These stats are calculated with minimal overhead by feeding in SchedulerOutput to the analytic config-based perf calculator at every iteration.

## How to use

Set `--enable-mfu-metrics` when launching the vLLM server to enable MFU logging.

`VLLM_DEBUG_MFU_METRICS=1` with `VLLM_LOGGING_LEVEL=DEBUG` is for debugging purposes for developers working on the metrics calculations.

## Purpose

To track the hardware performance over the course of vLLM serving.

## Examples

(for each example, please find the MFU stats at the end of each logged line)
(all examples were gathered on B200 GPUs)

### GPT-OSS 120B TP=8 BatchSize=256 Input=6K Output=3K NumBatch=1

**with `VLLM_MFU_LOGGING_LEVEL=1`**
```
INFO 12-15 17:38:18 [loggers.py:259] Engine 000: Avg prompt throughput: 3599.9 tokens/s, Avg generation throughput: 199.6 tokens/s, Running: 6 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.1%, Prefix cache hit rate: 0.0%, MFU: 6.6 TF/s/GPU 105.0 GB/s/GPU
INFO 12-15 17:38:28 [loggers.py:259] Engine 000: Avg prompt throughput: 150003.2 tokens/s, Avg generation throughput: 938.8 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 2.4%, Prefix cache hit rate: 0.0%, MFU: 259.9 TF/s/GPU 491.7 GB/s/GPU
INFO 12-15 17:38:38 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 18609.5 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 2.7%, Prefix cache hit rate: 0.0%, MFU: 32.6 TF/s/GPU 1138.1 GB/s/GPU
INFO 12-15 17:38:48 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 18479.1 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.0%, Prefix cache hit rate: 0.0%, MFU: 33.3 TF/s/GPU 1191.9 GB/s/GPU
INFO 12-15 17:38:58 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 18428.1 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.2%, Prefix cache hit rate: 0.0%, MFU: 34.2 TF/s/GPU 1249.8 GB/s/GPU
INFO 12-15 17:39:08 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 18201.0 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.5%, Prefix cache hit rate: 0.0%, MFU: 34.8 TF/s/GPU 1294.4 GB/s/GPU
INFO 12-15 17:39:18 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 2130.7 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, MFU: 4.1 TF/s/GPU 159.4 GB/s/GPU
INFO 12-15 17:39:28 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%
```

**with `VLLM_MFU_LOGGING_LEVEL=2`**
https://gist.github.com/SungMinCho/9ed5254e4bd1b3e5eb05a360df0dcc88

gives you the details behind the logged MFU stats such as
- per-component breakdown of reported flops/bytes
- breakdown of run context (i.e. input to the MFU calculator for that log duration e.g. prefill_num_tokens, etc)
- log duration, MFU calculation duration, MFU calculation overhead
- etc

## Some more examples with different parallelism setups

### GPT-OSS 120B TP=4 BatchSize=256 Input=6K Output=3K NumBatch=1

```
[0;36m(APIServer pid=2363162)[0;0m INFO 12-15 18:27:00 [loggers.py:259] Engine 000: Avg prompt throughput: 101389.0 tokens/s, Avg generation throughput: 77.4 tokens/s, Running: 170 reqs, Waiting: 86 reqs, GPU KV cache usage: 3.7%, Prefix cache hit rate: 0.0%, MFU: 349.4 TF/s/GPU 391.1 GB/s/GPU
[0;36m(APIServer pid=2363162)[0;0m INFO 12-15 18:27:10 [loggers.py:259] Engine 000: Avg prompt throughput: 51584.2 tokens/s, Avg generation throughput: 9711.6 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 5.3%, Prefix cache hit rate: 0.0%, MFU: 211.3 TF/s/GPU 1339.8 GB/s/GPU
[0;36m(APIServer pid=2363162)[0;0m INFO 12-15 18:27:20 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 15227.4 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 5.8%, Prefix cache hit rate: 0.0%, MFU: 54.0 TF/s/GPU 1870.2 GB/s/GPU
[0;36m(APIServer pid=2363162)[0;0m INFO 12-15 18:27:30 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 15176.8 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 6.3%, Prefix cache hit rate: 0.0%, MFU: 55.1 TF/s/GPU 1947.1 GB/s/GPU
[0;36m(APIServer pid=2363162)[0;0m INFO 12-15 18:27:40 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 15151.6 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 6.7%, Prefix cache hit rate: 0.0%, MFU: 56.3 TF/s/GPU 2026.6 GB/s/GPU
[0;36m(APIServer pid=2363162)[0;0m INFO 12-15 18:27:50 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 15074.7 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 7.2%, Prefix cache hit rate: 0.0%, MFU: 57.4 TF/s/GPU 2098.3 GB/s/GPU
```

### GPT-OSS 120B TP=1 BatchSize=256 Input=6K Output=3K NumBatch=1
```
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:32:12 [loggers.py:259] Engine 000: Avg prompt throughput: 600.0 tokens/s, Avg generation throughput: 204.3 tokens/s, Running: 1 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.3%, Prefix cache hit rate: 0.0%, MFU: 11.3 TF/s/GPU 772.8 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:32:22 [loggers.py:259] Engine 000: Avg prompt throughput: 29400.6 tokens/s, Avg generation throughput: 8.7 tokens/s, Running: 50 reqs, Waiting: 206 reqs, GPU KV cache usage: 9.1%, Prefix cache hit rate: 0.0%, MFU: 405.0 TF/s/GPU 269.3 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:32:32 [loggers.py:259] Engine 000: Avg prompt throughput: 37799.9 tokens/s, Avg generation throughput: 27.6 tokens/s, Running: 113 reqs, Waiting: 143 reqs, GPU KV cache usage: 17.4%, Prefix cache hit rate: 0.0%, MFU: 521.0 TF/s/GPU 344.9 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:32:42 [loggers.py:259] Engine 000: Avg prompt throughput: 37800.0 tokens/s, Avg generation throughput: 46.5 tokens/s, Running: 176 reqs, Waiting: 80 reqs, GPU KV cache usage: 25.6%, Prefix cache hit rate: 0.0%, MFU: 521.2 TF/s/GPU 349.3 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:32:52 [loggers.py:259] Engine 000: Avg prompt throughput: 37797.4 tokens/s, Avg generation throughput: 65.4 tokens/s, Running: 239 reqs, Waiting: 17 reqs, GPU KV cache usage: 33.8%, Prefix cache hit rate: 0.0%, MFU: 521.5 TF/s/GPU 353.6 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:33:02 [loggers.py:259] Engine 000: Avg prompt throughput: 10197.0 tokens/s, Avg generation throughput: 6142.2 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 34.7%, Prefix cache hit rate: 0.0%, MFU: 225.5 TF/s/GPU 2940.2 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:33:12 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6832.8 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 36.2%, Prefix cache hit rate: 0.0%, MFU: 95.6 TF/s/GPU 3238.4 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:33:22 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6807.8 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 37.6%, Prefix cache hit rate: 0.0%, MFU: 96.4 TF/s/GPU 3293.4 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:33:32 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6756.5 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 39.0%, Prefix cache hit rate: 0.0%, MFU: 96.7 TF/s/GPU 3334.6 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:33:42 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6755.3 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 40.5%, Prefix cache hit rate: 0.0%, MFU: 97.7 TF/s/GPU 3399.8 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:33:52 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6707.4 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 41.9%, Prefix cache hit rate: 0.0%, MFU: 98.1 TF/s/GPU 3440.7 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:34:02 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6603.0 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 43.3%, Prefix cache hit rate: 0.0%, MFU: 97.6 TF/s/GPU 3450.4 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:34:12 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6527.2 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 44.7%, Prefix cache hit rate: 0.0%, MFU: 97.4 TF/s/GPU 3472.5 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:34:22 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6576.8 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 46.1%, Prefix cache hit rate: 0.0%, MFU: 99.2 TF/s/GPU 3561.0 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:34:32 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6627.8 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 47.5%, Prefix cache hit rate: 0.0%, MFU: 100.9 TF/s/GPU 3651.6 GB/s/GPU
[0;36m(APIServer pid=2548211)[0;0m INFO 12-15 18:34:42 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 6551.6 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 48.9%, Prefix cache hit rate: 0.0%, MFU: 100.8 TF/s/GPU 3671.8 GB/s/GPU
```

### GPT-OSS 120B DP=2 TP=4 EP=8 BatchSize=256 Input=6K Output=3K NumBatch=1

Note: This is to show that it works well under multiple-engine scenarios (i.e. DP > 1)
```
INFO 12-15 18:51:34 [loggers.py:259] Engine 000: Avg prompt throughput: 600.0 tokens/s, Avg generation throughput: 98.9 tokens/s, Running: 1 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, MFU: 1.7 TF/s/GPU 50.0 GB/s/GPU
INFO 12-15 18:51:34 [loggers.py:259] Engine 001: Avg prompt throughput: 600.0 tokens/s, Avg generation throughput: 0.4 tokens/s, Running: 1 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, MFU: 0.4 TF/s/GPU 0.5 GB/s/GPU
INFO 12-15 18:51:44 [loggers.py:259] Engine 000: Avg prompt throughput: 62998.4 tokens/s, Avg generation throughput: 32.0 tokens/s, Running: 106 reqs, Waiting: 22 reqs, GPU KV cache usage: 2.5%, Prefix cache hit rate: 0.0%, MFU: 160.6 TF/s/GPU 106.9 GB/s/GPU
INFO 12-15 18:51:44 [loggers.py:259] Engine 001: Avg prompt throughput: 62998.7 tokens/s, Avg generation throughput: 32.0 tokens/s, Running: 106 reqs, Waiting: 22 reqs, GPU KV cache usage: 2.5%, Prefix cache hit rate: 0.0%, MFU: 160.6 TF/s/GPU 106.9 GB/s/GPU
INFO 12-15 18:51:54 [loggers.py:259] Engine 000: Avg prompt throughput: 13197.8 tokens/s, Avg generation throughput: 6962.0 tokens/s, Running: 128 reqs, Waiting: 0 reqs, GPU KV cache usage: 2.8%, Prefix cache hit rate: 0.0%, MFU: 51.6 TF/s/GPU 846.9 GB/s/GPU
INFO 12-15 18:51:54 [loggers.py:259] Engine 001: Avg prompt throughput: 13197.9 tokens/s, Avg generation throughput: 6962.0 tokens/s, Running: 128 reqs, Waiting: 0 reqs, GPU KV cache usage: 2.8%, Prefix cache hit rate: 0.0%, MFU: 51.6 TF/s/GPU 846.9 GB/s/GPU
INFO 12-15 18:52:04 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 9136.6 tokens/s, Running: 128 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.1%, Prefix cache hit rate: 0.0%, MFU: 24.5 TF/s/GPU 1136.6 GB/s/GPU
INFO 12-15 18:52:04 [loggers.py:259] Engine 001: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 9136.6 tokens/s, Running: 128 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.1%, Prefix cache hit rate: 0.0%, MFU: 24.5 TF/s/GPU 1136.6 GB/s/GPU
INFO 12-15 18:52:14 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 9147.9 tokens/s, Running: 128 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.4%, Prefix cache hit rate: 0.0%, MFU: 25.5 TF/s/GPU 1198.2 GB/s/GPU
INFO 12-15 18:52:14 [loggers.py:259] Engine 001: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 9147.9 tokens/s, Running: 128 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.4%, Prefix cache hit rate: 0.0%, MFU: 25.5 TF/s/GPU 1198.2 GB/s/GPU
INFO 12-15 18:52:24 [loggers.py:259] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 9073.2 tokens/s, Running: 128 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.7%, Prefix cache hit rate: 0.0%, MFU: 26.2 TF/s/GPU 1248.0 GB/s/GPU
INFO 12-15 18:52:24 [loggers.py:259] Engine 001: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 9073.3 tokens/s, Running: 128 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.7%, Prefix cache hit rate: 0.0%, MFU: 26.2 TF/s/GPU 1248.0 GB/s/GPU
```



## Test Plan

See **Examples** above.

## Test Result

See **Examples** above.

## Notes

This PR has been moved from https://github.com/vllm-project/vllm/pull/28859 due to code sync issues with Meta-internal codebase. See https://github.com/vllm-project/vllm/pull/28859 for some of the original discussions and reviews.

SungMinCho committed:
Add mfu stats logging

Signed-off-by: SungMinCho <tjdals4565@gmail.com>

SungMinCho commented:
Hi @markmc @zhuohan123 @bwasti   I moved https://github.com/vllm-project/vllm/pull/28859 into this PR to bypass internal codebase sync problems. Could you guys review this one last time and land? If anything, by default this functionality is turned off so it should be fairly safe to land. Thanks!

zhuohan123 reviewed (approved):
LGTM! I think the main todo is to add support for more complex attention types?

robertgshaw2-redhat commented on vllm/envs.py:
```diff
@@ -244,6 +244,7 @@
     VLLM_SHARED_EXPERTS_STREAM_TOKEN_THRESHOLD: int = 256
     VLLM_COMPILE_CACHE_SAVE_FORMAT: Literal["binary", "unpacked"] = "binary"
     VLLM_USE_V2_MODEL_RUNNER: bool = False
+    VLLM_MFU_LOGGING_LEVEL: int = 0  # 0: disabled, 1: enabled, 2: verbose
```
can you make this an engine arg instead?

SungMinCho committed:
Use ObservabilityConfig instead of ENVVAR to specify MFU logging level

Signed-off-by: SungMinCho <tjdals4565@gmail.com>

SungMinCho commented:
Hi @markmc @robertgshaw2-redhat I pushed a new commit reflecting all of your suggestions.

I removed VLLM_MFU_LOGGING_LEVEL from envs.py and added --mfu-metrics to observability config instead (thus made into engine args). (options: disabled (default), per-gpu, per-gpu-debug)

I also made PerfMetricsLogging receive log_fn just like every other sub-loggers, instead of the original log_parts an log_args in-place modification. Now we log MFU separately using the received log_fn.

I also separated the detail-logging into a separate logger.debug() statement (separate from the main-log_fn-based-logging described above), which is optionally enabled by --mfu-metrics=per-gpu-debug

## (modified) How To Use

By default MFU metrics logging is disabled.

To enable it, simply do
```
vllm serve ... --mfu-metrics=per-gpu
```

You will get something like

```
INFO 12-16 11:42:56 [loggers.py:257] Engine 000: Avg prompt throughput: 16199.9 tokens/s, Avg generation throughput: 178.3 tokens/s, Running: 27 reqs, Waiting: 75 reqs, GPU KV cache usage: 0.4%, Prefix cache hit rate: 0.0%
INFO 12-16 11:42:56 [perf.py:1178] MFU: 28.2 TF/s/GPU 131.0 GB/s/GPU
INFO 12-16 11:43:06 [loggers.py:257] Engine 000: Avg prompt throughput: 137367.0 tokens/s, Avg generation throughput: 3136.9 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 2.4%, Prefix cache hit rate: 0.0%
INFO 12-16 11:43:06 [perf.py:1178] MFU: 241.9 TF/s/GPU 585.9 GB/s/GPU
INFO 12-16 11:43:16 [loggers.py:257] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 18531.0 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 2.7%, Prefix cache hit rate: 0.0%
INFO 12-16 11:43:16 [perf.py:1178] MFU: 32.6 TF/s/GPU 1140.5 GB/s/GPU
INFO 12-16 11:43:26 [loggers.py:257] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 18376.1 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.0%, Prefix cache hit rate: 0.0%
INFO 12-16 11:43:26 [perf.py:1178] MFU: 33.3 TF/s/GPU 1192.1 GB/s/GPU
INFO 12-16 11:43:36 [loggers.py:257] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 18323.6 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.3%, Prefix cache hit rate: 0.0%
INFO 12-16 11:43:36 [perf.py:1178] MFU: 34.1 TF/s/GPU 1249.2 GB/s/GPU
INFO 12-16 11:43:46 [loggers.py:257] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 18222.3 tokens/s, Running: 255 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.5%, Prefix cache hit rate: 0.0%
INFO 12-16 11:43:46 [perf.py:1178] MFU: 34.9 TF/s/GPU 1302.3 GB/s/GPU
```

To enable it with debugging details simply do
```
VLLM_LOGGING_LEVEL=DEBUG vllm serve ... --mfu-metrics=per-gpu-debug
```

You will get something like

```
INFO 12-16 11:47:54 [v1/metrics/loggers.py:257] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 17859.8 tokens/s, Running: 256 reqs, Waiting: 0 reqs, GPU KV cache usage: 3.5%, Prefix cache hit rate: 0.0%
INFO 12-16 11:47:54 [v1/metrics/perf.py:1178] MFU: 34.0 TF/s/GPU 1262.0 GB/s/GPU
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200] MFU details: {
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   "prefill_reqs": 0,
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   "decode_reqs": 178688,
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   "num_batches": 698,
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   "context_breakdown": {
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "prefill_num_tokens": 0,
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "prefill_context_len": 0,
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "prefill_token_context_product": 0,
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "decode_num_tokens": 178688,
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "decode_context_len": 1511946176,
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "decode_token_context_product": 1511946176
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   },
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   "flops_breakdown": {
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.qkv_proj": "23.7TF",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.attn_qk": "55.7TF",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.attn_av": "55.7TF",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.out_proj": "19.0TF",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "ffn.routed_ffn": "160.1TF",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "unembed.unembed": "25.9TF"
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   },
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   "num_read_bytes_breakdown": {
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.qkv_input": "37.1GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.qkv_weight": "23.2GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.attn_input": "6973.6GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.out_input": "6.6GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.out_weight": "18.5GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "ffn.routed_up_gate_input": "148.2GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "ffn.routed_up_gate_weights": "3334.7GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "ffn.routed_silu_input": "37.1GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "ffn.routed_down_input": "18.5GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "ffn.routed_down_weights": "1667.4GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "unembed.input": "1.0GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "unembed.weight": "101.1GB"
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   },
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   "num_write_bytes_breakdown": {
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.qkv_output": "8.2GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.kv_cache": "0.8GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "attn.out_output": "37.1GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "ffn.routed_up_gate_output": "37.1GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "ffn.routed_silu_output": "18.5GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "ffn.routed_down_output": "148.2GB",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]     "unembed.output": "9.0GB"
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   },
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   "duration": "10.0s",
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200]   "mfu_calc_overhead": "0.9%"
DEBUG 12-16 11:47:54 [v1/metrics/perf.py:1200] }

```


As for Prometheus, yes I prefer to do it as a follow up after landing this.

Thanks for all the feedbacks!

markmc commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,1241 @@
+            total.update(prefixed)
+        return total
+
+    def get_perf_stats_per_gpu(self, ctx: ExecutionContext) -> PerfStats:
```
This appears unused and should (presumably) be removed?

markmc committed:
[MFU] Move request counts into ExecutionContext

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[MFU] Delete unused get_perf_stats_per_gpu()

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[MFU] Remove unused PerfStats.__add__()

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[MFU] Re-instate test suite

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[MFU] Refactor debug logging

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[MFU] Change CLI arg to --enable-mfu-metrics

And use VLLM_DEBUG_MFU_METRICS to enable debugging.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,1235 @@
+            (self.total_read_bytes_per_gpu + self.total_write_bytes_per_gpu)
+            / delta_time_per_gpu
+            / 1e9
+        )
```
@SungMinCho what's your take on this comment?

markmc reviewed (changes requested):
I'm keen to merge this without more back-and-forth on the CLI arg vs env var, etc. etc. - I'd imagine you don't have super strong opinions on this, so I'll push some updates with my take to your branch ... feel free to revert anything you disagree with!

markmc commented:
Rebased to pick up #30878

SungMinCho committed:
Add Engine visibility to MFU logs & Fix PP duration

Signed-off-by: SungMinCho <tjdals4565@gmail.com>

SungMinCho commented:
@markmc JFYI land is currently blocked due to some build fail issue in internal trunk which is irrelevant to this PR... We might have to wait until the oncall resolves that issue... It's really frustrating that I can't merge this PR to OSS without having to sync to internal stack. @zhuohan123 is there any possible bypass?

SungMinCho commented:
Assuming @zhuohan123 can bypass internal sync and click merge button,

the current CI still seems to have 2 failures.

<img width="311" height="101" alt="image" src="https://github.com/user-attachments/assets/84de93ec-0596-456a-9747-7864c453f1f3" />


- Neuron Test (bash .buildkite/scripts/hardware_ci/run-neuron-test.sh)
  - https://buildkite.com/vllm/ci/builds/44008/steps/canvas?sid=019b2e26-d9d1-40c6-b487-4fe3e5161150
  - `bash: .buildkite/scripts/hardware_ci/run-neuron-test.sh: No such file or directory`
- Ascend NPU Test
  - https://buildkite.com/vllm/ci/builds/44008/steps/canvas?sid=019b2e26-d9d3-4270-9667-ab32342122dc
  -  `requests.exceptions.ConnectionError: HTTPSConnectionPool(host='www.modelscope.cn', port=443): Max retries exceeded with url: /api/v1/models/Qwen/Qwen3-0.6B (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0xfffd0a4fe050>: Failed to resolve 'www.modelscope.cn' ([Errno -3] Temporary failure in name resolution)"))`


Do we know if this is a known issue at the moment? (cc @markmc)

SungMinCho commented:
Thanks @zhuohan123 for merging!
